### PR TITLE
Custom Kinetic Accelerators

### PR DESCRIPTION
**Pull request by BurgerLUA**
# Overview
This is honestly one of my biggest and most ambitious projects. I hope people are happy with this.

Custom kinetic accelerators are special, customizable mining weapons that are meant to give a sense of progression, and a sense of pride and accomplishment while mining. Each custom KA is made up of 3 or 4 parts:

- The Frame (5 to choose from)
- The Cell (5 to choose from, +1 traitor)
- The Barrel (5 to choose from, +1 traitor)
- The Upgrade Chip (7 to choose from, +1 traitor)

The sprites change with each addon, they are truly dynamic and there are so many combinations of parts that a miner can have. You can have several different builds for each desired mining style, it's quite a robust system.

You can have a KA that shoots slowly, but delivers high-penetrating shots that go through several walls of rock. You can have a KA that shoots REALLY fast, but precisely destroys 1 rock at a time. You can have an absolute canon of a beast, that destroys rocks in a 3 tile radius around it.

The parts can be obtained via research, via abandoned crates, or via RNG in the cargo warehouse. There are some custom KAs that spawn on the raider's ship, on the mercs ship, and a laser one on the merc's headquarters. That laser one can also be bought from the traitor uplink as well.

ERT get a class 4 KA for use in special operations, if they choose to use it.

An overview of sprites here: https://imgur.com/a/x9VvVVa

### CODE REVIEW

**Comment by BotBOREALIS** [comment]
Icons with diff:
<details><summary>icons/obj/kinetic_accelerators.dmi</summary>

Key | Old | New | Status
--- | --- | --- | ---
frame01||frame01|Created
frame02||frame02|Created
frame03||frame03|Created
frame04||frame04|Created
frame05||frame05|Created
barrel01||barrel01|Created
barrel02||barrel02|Created
barrel03||barrel03|Created
barrel04||barrel04|Created
barrel05||barrel05|Created
cell01||cell01|Created
cell02||cell02|Created
cell03||cell03|Created
cell04||cell04|Created
cell05||cell05|Created
upgrade_chip||upgrade_chip|Created
|||Created
barrel_laser||barrel_laser|Created
upgrade_chip_illegal||upgrade_chip_illegal|Created
cell_illegal||cell_illegal|Created
kineticgun_lh||kineticgun_lh|Created
kineticgun_lh[1]||![kineticgun_lh[1]](http://images.aurorastation.org/images/86cc280ab9374687bc4811e5fec042e4.png)|Created
kineticgun_lh[2]||![kineticgun_lh[2]](http://images.aurorastation.org/images/e18c73b0687eb590c0c9a02ac3a3828b.png)|Created
kineticgun_lh[3]||![kineticgun_lh[3]](http://images.aurorastation.org/images/6889f7c3ac52a940f43ac0d55b5b8b05.png)|Created
kineticgun_rh||kineticgun_rh|Created
kineticgun_rh[1]||![kineticgun_rh[1]](http://images.aurorastation.org/images/79c85164540871ce93f972ce03554bc3.png)|Created
kineticgun_rh[2]||![kineticgun_rh[2]](http://images.aurorastation.org/images/6a37caad89bcf71e1c3be63fc2c8a406.png)|Created
kineticgun_rh[3]||![kineticgun_rh[3]](http://images.aurorastation.org/images/5892e4be6addc443329536182991d46b.png)|Created
</details>

**Comment by Pacmandevil** [comment]
inb4 someone complains because the stock KPA winds up being like 1% less effective

**Comment by OneOneThreeEight** [comment]
wouldn't that be weird, that an un-upgraded device isn't as effective as an upgraded one? nah, totally worth the outrage

File: code/modules/custom_ka/barrels.dm [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	cost_increase = 1
+	cell_increase = 0
+	capacity_increase = -1
+	fire_sound = 'sound/weapons/Kenetic_accel.ogg'
+	projectile_type = /obj/item/projectile/kinetic
+	origin_tech = list(TECH_MATERIAL = 2,TECH_ENGINEERING = 2,TECH_MAGNET = 2)
+
+/obj/item/custom_ka_upgrade/barrels/barrel02
+	name = "professional core KA power converter"
+	build_name = "core XXC 'Profesional'"
```
Comment (OneOneThreeEight) [inline code/modules/custom_ka/barrels.dm]: Misspell of 'professional.' Not to be nitpicky.

File: code/modules/custom_ka/frames.dm [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+/obj/item/projectile/kinetic
+	name = "kinetic force"
+	icon_state = null
+	damage = 15
+	damage_type = BRUTE
+	check_armour = "bomb"
+	range = 5
+
+	var/pressure_decrease = 0.25
+	var/list/hit_overlays = list()
```
Comment (Lohikar) [inline code/modules/custom_ka/frames.dm]: This list is never written to, only read.

**Comment by Alberyk** [comment]
Some issues:
-the illegal ka is pretty much a laser weapon, it is pointless to give them to the mercs at round start when they already have access to laser carbines
-the names are way too long and they clash with the other gun names

**Comment by Alberyk** [comment]
Also, what is exactly being changed on the maps?

**Comment by BurgerLUA** [comment]
I added a single wrench to the main level in mining, as wrenches are required to modify custom KAs.
I also added 2 random, custom KAs on the raider shuttle, and 3 random, custom KAs on the merc shuttle, and 3 level 4 custom KAs in the ERT specialist area.

Because mapmerge is mapmerge, it did a lot of file-related changes.

**Comment by Alberyk** [comment]
Pretty sure that mining already has a toolbox in the mining bay, and I don't see the need for mercs or raiders to have access to those, they already have weapons and mining drills. And why does even ert needs them?

**Comment by BurgerLUA** [comment]
Reason: I want people to try them out.

Excuse: ERT only have 2 things of mining equipment, and it's the hardsuit mounted drill. I just think it would be pretty neat for ERT to get non-plebeian equipment if a dig-related situation arises. Raiders would likely have access to them given how they steal, and given how the KAs they steal are randomized. It would be possible for them to randomize a KA that doesn't work. I'll probably get rid of the mercs KA since they already have diamond drills.

**Comment by BurgerLUA** [comment]
I can't perform a mapmerge on the centcomm level because map merge 1 is not smart enough to do it, for some reason. Waiting until there's a fix to this problem until I attempt merging.

**Comment by BurgerLUA** [comment]
Fixed the merge conflict. Centcomm and map merge fixer do not like behaving, so centcomm is now left unchanged, meaning that ERT and Mercs no longer get custom KAs.

File: code/game/objects/items/weapons/manuals.dm [hunk 3]
```diff
@@ -1367,4 +1367,84 @@
 				Remember; once you have been afflicted with brain trauma you are over four times as likely to suffer it again! Keep your patient's best interests at heart.
 				</body>
 			</html>
-			"}
\ No newline at end of file
+			"}
+
+/obj/item/weapon/book/manual/brainwashing
```
Comment (LordFowl) [inline code/game/objects/items/weapons/manuals.dm]: What does brainwashing have to do with pKAs?

File: code/modules/custom_ka/barrels.dm [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	cell_increase = 0
+	capacity_increase = -4
+	fire_sound = 'sound/weapons/pulse.ogg'
+	projectile_type = /obj/item/projectile/kinetic
+	origin_tech = list(TECH_MATERIAL = 5,TECH_ENGINEERING = 5,TECH_MAGNET = 5)
+
+/obj/item/custom_ka_upgrade/barrels/barrel05
+	name = "experimental core KA power converter"
+	build_name = "'Experimental'"
+	desc = "A very experimental kinetic accelerator energy converter. Not much is known about this thing, other than it kicks like a muel and stings like an e-sword."
```
Comment (LordFowl) [inline code/modules/custom_ka/barrels.dm]: Mule. And I've never heard of e-swords being notable for their stinginess.

File: code/modules/custom_ka/barrels.dm [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+	cell_increase = 0
+	capacity_increase = -5
+	fire_sound = 'sound/weapons/resonator_blast.ogg'
+	projectile_type = /obj/item/projectile/kinetic
+	origin_tech = list(TECH_MATERIAL = 6,TECH_ENGINEERING = 6,TECH_MAGNET = 6)
+
+/obj/item/custom_ka_upgrade/barrels/illegal
+	name = "laser KA power converter"
+	build_name = "'Syndicate''"
+	desc = "A laser cystal ripped from a laser rifle and repurposed for kinetic accelerator assemblies."
```
Comment (LordFowl) [inline code/modules/custom_ka/barrels.dm]: Crystal.

File: code/modules/custom_ka/core.dm [hunk 6]
```diff
@@ -0,0 +1,497 @@
...
+	if(installed_upgrade_chip)
+		to_chat(user,"It is equipped with \the [installed_barrel], \the [installed_cell], and \the [installed_upgrade_chip].")
+	else if(installed_barrel)
+		to_chat(user,"It is equipped with \the [installed_barrel] and \the [installed_cell]. It has space for an upgrade chip.")
+	else if(installed_cell)
+		to_chat(user,"It is equipped with \the [installed_cell]. The assembly lacks a barrel installation.")
+
+	if(installed_barrel)
+		if(custom_name)
+			to_chat(user,"[custom_name] is written crudely in pen across the side, with the offical designation \"[official_name]\" is etched neatly on the side.")
```
Comment (LordFowl) [inline code/modules/custom_ka/core.dm]: Crude grammar. `While the official designation X is etched neatly...` or `With the official designation X etched neatly...`

File: code/modules/custom_ka/upgrade_chips.dm [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+/obj/item/custom_ka_upgrade/upgrade_chips/recoil
+	name = "upgrade chip - recoil reduction"
+	desc = "Reduces recoil, increases accuracy, reduces rate of fire"
+	icon_state = "upgrade_chip"
+	recoil_increase = -5
+	firedelay_increase = 0.25 SECONDS
+
+/obj/item/custom_ka_upgrade/upgrade_chips/focusing
+	name = "upgrade chip - focusing"
+	desc = "Increases damage and accuracy, however reduces the range and reduce aoe capabilities."
```
Comment (LordFowl) [inline code/modules/custom_ka/upgrade_chips.dm]: Reduces.

File: code/modules/mining/machine_rigpress.dm [hunk 8]
```diff
@@ -19,8 +19,10 @@
 /obj/machinery/mineral/rigpress/attackby(obj/item/W, mob/user)
 	if(!pressing)
 		var/outcome_path
-		var/list/kinetic_mods = list()
-		var/kineticaccelerator
+
+		//var/list/kinetic_mods = list()
```
Comment (LordFowl) [inline code/modules/mining/machine_rigpress.dm]: Remove, don't comment out.

File: code/modules/mining/machine_rigpress.dm [hunk 9]
```diff
@@ -33,12 +35,13 @@
 		if(istype(W, /obj/item/weapon/pickaxe/drill))
 			outcome_path = /obj/item/rig_module/device/basicdrill
 
-		if(istype(W, /obj/item/weapon/gun/energy/kinetic_accelerator))
-			outcome_path = /obj/item/rig_module/mounted/kinetic_accelerator
-			var/obj/item/weapon/gun/energy/kinetic_accelerator/KA = W
-			kineticaccelerator = 1
-			for(var/obj/item/borg/upgrade/modkit/kmod in KA.modkits)
-				kinetic_mods += kmod
+		//TODO: MOUNTED CUSTOM KAS
```
Comment (LordFowl) [inline code/modules/mining/machine_rigpress.dm]: Likewise. Too many abandoned todos in this codebase to count.

File: code/modules/mining/machine_rigpress.dm [hunk 10]
```diff
@@ -61,13 +64,17 @@
...
 				ping( "\The [src] pings, \"Module successfuly produced!\"" )
-				if(kineticaccelerator)
-					var/obj/item/rig_module/mounted/kinetic_accelerator/KA = new /obj/item/rig_module/mounted/kinetic_accelerator(src.loc)
-					var/obj/item/weapon/gun/energy/kinetic_accelerator/cyborg/KGUN = KA.gun
-					for(var/obj/item/borg/upgrade/modkit/kmod in kinetic_mods)
-						KGUN.modkits += kmod
-				else
-					new outcome_path(src.loc)
+
+				//TODO: MOUNTED CUSTOM KAS
```
Comment (LordFowl) [inline code/modules/mining/machine_rigpress.dm]: Likewise.

File: html/changelogs/burgerbb-ka.yml [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+delete-after: True
+
+# Any changes you've made.  See valid prefix list above.
+# INDENT WITH TWO SPACES.  NOT TABS.  SPACES.
+# SCREW THIS UP AND IT WON'T WORK.
+# Also, all entries are changed into a single [] after a master changelog generation. Just remove the brackets when you add new entries.
+# Please surround your changes in  double quotes ("), as certain characters otherwise screws up compiling. The quotes will not show up in the changelog.
+changes: 
+  - rscadd: "Removed basic kinetic accelerators from code. Replaced them with customizable kinetic accelerators with a robust array of customization. Weaker custom kinetic accelerators can be purchased from the mining vendor or found randomly in cargo, while the stronger variants can be researched and produced by science."
+  - rscadd: "Added traitor kinetic accelerators to uplink, as well as one in the merc headquarters. These ones shoot laser beams, and can accept custom accelerator parts."
```
Comment (LordFowl) [inline html/changelogs/burgerbb-ka.yml]: This is no longer true.

File: code/game/objects/items/weapons/manuals.dm (hunk 3 above)
Comment (BurgerLUA) [reply code/game/objects/items/weapons/manuals.dm]: It was leftover from an poorly done devmerge on my part.

**Comment by BurgerLUA** [comment]
@LordFowl The level 1 kinetic accelerator was accidentally made powerful due to an oversight in code and an attempt to make all modules "equal but different". It fires significantly slower than the original KA, however. I also hardcapped the damage per projectile to living targets to 30 to avoid powercreep.

Recoil based reloading takes away recoil from the gun because it converts that recoil into kinetic energy, effectively removing/reducing it.

As for the snowflake implementation, there are plans to merge the two systems together in a future PR

**Comment by LordFowl** [comment]
I'm not really keen on the "I'll fix it later" conditional, but I will in this incident defer to an additional judgement.

**Comment by LordFowl** [comment]
I personally don't think merging a "faulty" (in the sense that it is redundant, not necessarily poor) system under the promise of "fixing" (e.g. merging the two systems to not be redundant) later is a good practice.

**Comment by BurgerLUA** [comment]
@LordFowl I am making a whole bunch of new sprites and whatnot for a future update. I will keep that feedback in mind, as well as future feedback on the current KAs, when implementing the changes,

**Pull request merged**